source/common/upstream/cluster_manager_impl.h

```diff
@@ -332,10 +338,115 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+        : parent_(parent), odcds_(std::move(odcds)) {
+      ASSERT(odcds_ != nullptr);
+    }
+
+    ClusterDiscoveryCallbackHandlePtr
+    requestOnDemandClusterDiscovery(const std::string& name,
+                                    ClusterDiscoveryCallbackSharedPtr callback,
+                                    std::chrono::milliseconds timeout) override {
+      return parent_.requestOnDemandClusterDiscovery(odcds_, name, std::move(callback), timeout);
+    }
+
+  private:
+    ClusterManagerImpl& parent_;
+    OdCdsApiSharedPtr odcds_;
+  };
+
   virtual void postThreadLocalClusterUpdate(ClusterManagerCluster& cm_cluster,
                                             ThreadLocalClusterUpdateParams&& params);
 
+  /**
+   * Notifies cluster discovery managers in each worker thread that
+   * the discovery process for the cluster with a passed name has
+   * timed out.
+   */
+  void notifyExpiredDiscovery(const std::string& name);
+
 private:
+  struct ThreadLocalClusterManagerImpl;
+
+  /** A thread-local on-demand cluster discovery manager. It takes care of invoking the discovery
+   *  callbacks in the event of a finished discovery. It does it by installing a cluster lifecycle
+   *  callback that invokes the discovery callbacks when a matching cluster just got added.
+   */
+  class ClusterDiscoveryManager : Logger::Loggable<Logger::Id::upstream> {
+  public:
+    ClusterDiscoveryManager(ThreadLocalClusterManagerImpl& parent);
+
+    /**
+     * Invoke the callbacks for the given cluster name. The discovery status is passed to the
+     * callbacks. After invoking the callbacks, they are dropped from the manager.
+     */
+    void processClusterName(const std::string& name, ClusterDiscoveryStatus cluster_status);
+
+    /**
+     * A pair containing a discovery handle and information whether a discovery for a given cluster
+     * was already requested in this thread.
+     */
+    struct Pair {
+      ClusterDiscoveryCallbackHandlePtr handle_ptr_;
+      bool discovery_in_progress_;
+    };
+
+    /**
+     * Adds the discovery callback. Returns a handle and a boolean indicating whether this worker
+     * thread has already requested the discovery of a cluster with a given name.
+     */
+    Pair addCallback(const std::string& name, const ClusterDiscoveryCallbackSharedPtr& callback);
+
+  private:
+    using CallbackList = std::list<ClusterDiscoveryCallbackWeakPtr>;
+    using CallbackListIterator = CallbackList::iterator;
+
+    /**
+     * An implementation of discovery handle. Destroy it to drop the callback from the discovery
+     * manager. It won't stop the discovery process, though.
+     */
+    class ClusterDiscoveryCallbackHandleImpl : public ClusterDiscoveryCallbackHandle {
+    public:
+      ClusterDiscoveryCallbackHandleImpl(ClusterDiscoveryManager& parent, std::string name,
+                                         CallbackListIterator it)
+          : parent_(parent), name_(std::move(name)), it_(std::move(it)) {}
```

Comment:
A nit: consider an rval reference here for `name`?

Reply:
At the only call place of `ClusterDiscoveryCallbackHandleImpl` I pass a `const std::string&`, so I'd need to make a copy anyway.